autochain/models/chat_openai.py

```diff
@@ -230,3 +230,129 @@ def _create_llm_result(self, response: Mapping[str, Any]) -> LLMResult:
+    openai_api_base: Optional[str] = None
+    openai_api_type: str = "azure"
+    openai_api_version: str = "2023-05-15"
+    deployment_name: str = ""
+    request_timeout: Optional[Union[float, Tuple[float, float]]] = None
+    """Timeout for requests to OpenAI completion API. Default is 600 seconds."""
+    max_retries: int = 6
+    """Maximum number of retries to make when generating."""
+    # TODO: support streaming
+    # streaming: bool = False
+    # """Whether to stream the results or not."""
+    # n: int = 1
+    # """Number of chat completions to generate for each prompt."""
+    max_tokens: Optional[int] = None
+    """Maximum number of tokens to generate."""
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.ignore
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        """Validate that api key and python package exists in environment."""
+        openai_api_key = os.environ["OPENAI_API_KEY"]
+        openai_api_base = os.environ["OPENAI_API_BASE"]
+        try:
+            import openai
+
+        except ImportError:
+            raise ValueError(
+                "Could not import openai python package. "
+                "Please install it with `pip install openai`."
+            )
+        openai.api_key = openai_api_key
+        openai.api_base = openai_api_base
+        try:
+            values["client"] = openai.ChatCompletion
+        except AttributeError:
+            raise ValueError(
+                "`openai` has no `ChatCompletion` attribute, this is likely "
+                "due to an old version of the openai package. Try upgrading it "
+                "with `pip install --upgrade openai`."
+            )
+        # if values["n"] < 1:
+        #     raise ValueError("n must be at least 1.")
+        # if values["n"] > 1 and values["streaming"]:
+        #     raise ValueError("n must be 1 when streaming.")
+        return values
+
+    def generate(
+        self,
+        messages: List[BaseMessage],
+        functions: Optional[List[Tool]] = None,
+        stop: Optional[List[str]] = None,
+    ) -> LLMResult:
+        message_dicts, function_dicts, params = self._create_message_dicts(
+            messages, functions, stop
+        )
+
+        generation_param = {
+            "messages": message_dicts,
+            **params,
+        }
+        if len(function_dicts) > 0:
+            generation_param["functions"] = function_dicts
+
+        response = self.generate_with_retry(**generation_param)
+        return self._create_llm_result(response)
+
+    def _create_message_dicts(
+        self,
+        messages: List[BaseMessage],
+        tools: Optional[List[Tool]],
+        stop: Optional[List[str]],
+    ) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]], Dict[str, Any]]:
+        params: Dict[str, Any] = {**{"model": self.model_name, "engine": self.deployment_name}, **self._default_params}
```

Comment:
to reuse the same model, update `params` if self.azure_deployment_name is not None and self.api_type == "azure"